Sources/SystemInternals/Mocking.swift

```diff
@@ -0,0 +1,164 @@
+/*
+ This source file is part of the Swift System open source project
+
+ Copyright (c) 2020 Apple Inc. and the Swift System project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+*/
+
+// Syscall mocking support.
+//
+// NOTE: This is currently the bare minimum needed for System's testing purposes, though we do
+// eventually want to expose some solution to users.
+//
+// Mocking is contextual, accessible through MockingDriver.withMockingEnabled. Mocking
+// state, including whether it is enabled, is stored in thread-local storage. Mocking is only
+// enabled in testing builds of System currently, to minimize runtime overhead of release builds.
+//
+
+public struct Trace {
```

Comment:
I think the entire mocking infrastructure ought to be compiled out of the codebase when we aren't building for testing -- so this entire file should be inside a `#if ENABLE_MOCKING` block, or it should live in a special target that's only ever imported in the test targets.

Reply:
`@testable` would be a good way to hide these.

But I still think these should be entirely compiled out from build products that external code links with. (I.e., the mock* functions and the if conditions within the syscall shims must not be compiled into production libraries.) 

We are forced to do invasive testing here because we lack good alternatives -- but potentially leaving traces of the testing infrastructure in production code is a hazard I don't like.

Reply:
Internal code that is never called can be dropped by the compiler (and even the linker), unless there is something preventing the optimization. I'll do a bit of both.

Reply:
I don't think it's a good idea to rely on the optimizer to eliminate these -- the mocking entry points introduce a level of userspace dynamism that feels inappropriate for syscalls.